sets.c

```diff
@@ -0,0 +1,305 @@
+                    /* fill up array with range */
+                    for (size_t i = 0; i <= last_num; i++)
+                        valarr[vaidx++] = i;
+                    continue;
+                }
+                if (toknum <= last_num || toknum >= maxsz) {
+                    fail_reason = "invalid range";
+                    goto error;
+                }
+
+                /* pre-set the new size filled and grow array if necessary */
+                if ((arr_szfilled += (toknum - last_num)) > arr_maxsz)
+                    while (arr_szfilled > arr_maxsz) {
+                        if (!inc_arr_sz(&valarr, &arr_maxsz, maxsz, &fail_reason))
+                            goto error;
+                    }
+
+                /* fill up array with range */
+                for (size_t i = last_num+1; i <= toknum; i++)
+                    valarr[vaidx++] = i;
+            } else if (last_type == NUMBER_TOK) {
+                /* sanity check */
+                fail_reason = "impossible condition (last_type == NUMBER_TOK)";
+                goto error;
+            } else {
+                valarr[vaidx++] = toknum;
+                arr_szfilled++;
+            }
+            last_num  = toknum;
+            last_type = NUMBER_TOK;
+            if (!got_num)
+                got_num = true;
+            /* move token position to last number found (needed for 2+ digit nums) */
+            tok = tmpnum_end-1;
+        } else {
+            fail_reason = "bad token";
+            goto error;
+        }
+    }
+
+    /* check for {n .. end} range */
+    if (last_type == RANGE_TOK) {
+        if (!got_num) {
+            fail_reason = "invalid range";
+            goto error;
+        }
+        if (last_num >= maxsz) {
+            fail_reason = "n..end range OOB (Out Of Bounds)";
+            goto error;
+        }
+        if ((arr_szfilled += (maxsz-1 - last_num)) > arr_maxsz)
+            while (arr_szfilled > arr_maxsz) {
+                if (!inc_arr_sz(&valarr, &arr_maxsz, maxsz, &fail_reason))
+                    goto error;
+            }
+        /* fill up array with range */
+        for (size_t i = last_num+1; i < maxsz; i++)
+            valarr[vaidx++] = i;
+    }
+
+    /* consider empty sets invalid */
+    if (arr_szfilled == 0) {
+        fail_reason = "empty set";
+        goto error;
+    }
+
+    /* sort the value array */
+    qsort(valarr, arr_szfilled, sizeof(valarr[0]), _size_t_cmp);
+
+    /* check if there are any duplicates */
+    for (size_t i = 0; i+1 < arr_szfilled; i++)
+        if (valarr[i] == valarr[i+1]) {
+            fail_reason = "duplicate element";
+            goto error;
+        }
+
+    /* check range (for individual entries, faster than checking each one) */
+    if (valarr[arr_szfilled-1] >= maxsz) {
+        fail_reason = "OOB (Out Of Bounds) element(s)";
+        goto error;
+    }
+
+    size_t *tmp_vaptr;
+    if ((tmp_vaptr = realloc(valarr, arr_szfilled)) == NULL) {
```

Comment:
Here is missing a `* sizeof(size_t)`, which explains the absurd behaviour.

Reply:
Good catch, pushed. Thanks